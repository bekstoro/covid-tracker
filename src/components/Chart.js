import React from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import { IconButton } from '@material-ui/core';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';

import { updateParamsDate } from '../store/params.slice';

import ChartCountry from './ChartCountry';
import ChartDate from './ChartDate';
import ChartProvince from './ChartProvince';
import Title from './Title';

const useStyles = makeStyles(theme => ({
  icon: {
    marginRight: theme.spacing(1),
  },
}));

export default function Chart({ params }) {
  const classes = useStyles();
  const dispatch = useDispatch();
  
  if (params.date) {
    return (
      <>
        <Title>
          <IconButton color="primary" className={classes.icon} onClick={() => dispatch(updateParamsDate(''))}>
            <ArrowBackIcon/>
          </IconButton>
          Results for {params.date} with total {params.totalCases} cases ({params.totalDeaths} deaths)
        </Title>
        <ChartDate/>
      </>
    );
  }
  
  return (
    <>
      <Title>Results</Title>
      {params.province ? <ChartProvince/> : <ChartCountry/>}
    </>
  );
}

Chart.propTypes = {
  params: PropTypes.shape({
    country: PropTypes.string,
    date: PropTypes.string,
    period: PropTypes.number,
    province: PropTypes.string,
    totalCases: PropTypes.number,
    totalDeaths: PropTypes.number,
  }),
};
