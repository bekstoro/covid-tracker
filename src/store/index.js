import { configureStore } from '@reduxjs/toolkit';

import paramsReducer from './params.slice';
import provincesReducer from './provinces.slice';
import reportsReducer from './reports.slice';

export const store = configureStore({
  reducer: {
    params: paramsReducer,
    provinces: provincesReducer,
    reports: reportsReducer,
  },
});
