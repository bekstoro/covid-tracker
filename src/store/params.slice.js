import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  country: '',
  date: '',
  period: 0,
  province: '',
  totalCases: 0,
  totalDeaths: 0,
};

export const paramsSlice = createSlice({
  name: 'params',
  initialState,
  reducers: {
    update: (state, action) => {
      const {country, period, province} = action.payload;
      state.country = country;
      state.date = '';
      state.period = period;
      state.province = province;
      state.totalCases = 0;
      state.totalDeaths = 0;
    },
    updateDate: (state, action) => {
      state.date = action.payload;
    },
    updateTotal: (state, action) => {
      state.totalCases = action.payload.totalCases;
      state.totalDeaths = action.payload.totalDeaths;
    },
  },
});

export const { update: updateParams, updateDate: updateParamsDate, updateTotal: updateParamsTotal } = paramsSlice.actions;

export const selectParams = (state) => state.params;

export default paramsSlice.reducer;
