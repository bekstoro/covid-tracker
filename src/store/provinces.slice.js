import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { fetchProvinces } from '../api/provinces.api'

// actions
export const fetchProvincesAction = createAsyncThunk(
  'provinces/fetch',
  async () => {
    return fetchProvinces();
  }
);

// reducer
const initialState = {
  error: undefined,
  loading: false,
  provinces: [],
};
export const provincesSlice = createSlice({
  name: 'provinces',
  initialState,
  extraReducers: builder => {
    builder
      .addCase(fetchProvincesAction.pending, state => {
        state.loading = true;
      })
      .addCase(fetchProvincesAction.fulfilled, (state, action) => {
        state.loading = false;
        state.provinces = action.payload?.map(curr => ({ label: curr, value: curr }));
      })
      .addCase(fetchProvincesAction.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      });
  },
});

// selectors
export const selectProvinces = state => state.provinces.provinces;

export default provincesSlice.reducer;
