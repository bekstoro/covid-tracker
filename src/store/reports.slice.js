import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { fetchReportsByCountry, fetchReportsByCountryAndDate, fetchReportsByProvince } from '../api/reports.api';

// actions
export const fetchReportsByCountryAction = createAsyncThunk(
  'reports/fetchByCountry',
  async ({ country, period }) => {
    return fetchReportsByCountry(country, period);
  },
);

export const fetchReportsByCountryAndDateAction = createAsyncThunk(
  'reports/fetchByCountryAndDate',
  async ({ country, period }) => {
    return fetchReportsByCountryAndDate(country, period);
  },
);

export const fetchReportsByProvinceAction = createAsyncThunk(
  'reports/fetchByProvince',
  async ({ province, period }) => {
    return fetchReportsByProvince(province, period);
  },
);

// reducer
const initialState = {
  loading: false,
  countryProvinces: [],
  countries: [],
  provinces: [],
};
export const reportsSlice = createSlice({
  name: 'reports',
  initialState,
  extraReducers: builder => {
    builder
      .addCase(fetchReportsByCountryAction.pending, state => {
        state.loading = true;
      })
      .addCase(fetchReportsByCountryAction.fulfilled, (state, action) => {
        state.loading = false;
        state.countries = action.payload;
      })
      .addCase(fetchReportsByCountryAction.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })
      .addCase(fetchReportsByCountryAndDateAction.pending, state => {
        state.loading = true;
      })
      .addCase(fetchReportsByCountryAndDateAction.fulfilled, (state, action) => {
        state.loading = false;
        state.countryProvinces = action.payload;
      })
      .addCase(fetchReportsByCountryAndDateAction.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })
      .addCase(fetchReportsByProvinceAction.pending, state => {
        state.loading = true;
      })
      .addCase(fetchReportsByProvinceAction.fulfilled, (state, action) => {
        state.loading = false;
        state.provinces = action.payload;
      })
      .addCase(fetchReportsByProvinceAction.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      });
  },
});

// selectors
export const selectLoading = state => state.reports.loading;

export const selectReportsByCountry = state => state.reports.countries.timeline;

export const selectReportsByCountryAndDate = state => state.reports.countryProvinces.states;

export const selectReportsByProvince = state => state.reports.provinces;

export default reportsSlice.reducer;
