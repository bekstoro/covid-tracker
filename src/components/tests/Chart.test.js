import React from 'react';
import { Provider } from 'react-redux';

import '@testing-library/jest-dom';
import { render } from '@testing-library/react';

import { store } from '../../store';
import Chart from '../Chart';

let component;
const mockProps = {
  params: {
    country: 'USA',
    date: '',
    period: 7,
    province: '',
    totalCases: 0,
    totalDeaths: 0,
  }
}

describe('Chart component', () => {
  beforeEach(() => {
    component = render(
      <Provider store={store}>
        <Chart {...mockProps} />
      </Provider>
    );
  });
  
  it('should render correctly', () => {
    expect(component.asFragment()).toMatchSnapshot();
  });
});
