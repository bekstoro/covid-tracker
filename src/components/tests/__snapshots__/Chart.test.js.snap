// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component should render correctly 1`] = `
<DocumentFragment>
  <h2
    class="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
  >
    Results
  </h2>
  <div
    class="makeStyles-root-2"
  >
    <div
      class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
